overlord/README.md

```diff
@@ -0,0 +1,80 @@
+
+ During execution, a `Task` goes through a series of statuses. These are represented by state.Status and will finish in a ready status of either `DoneStatus, UndoneStatus, ErrorStatus` or `HoldStatus`.
+
+ If errors are encountered, the `TaskRunner `will normally try to recursively execute the undo logic of any previously depended-upon `Task`s with the exception of the `Task` that generated the error. It is instead expected that any desired undo should be part of its error paths.
+
+Different `Change`s and independent `Task`s are normally executed concurrently.
+
+`Task`s and `State` locking and consistency
+--------------------------------------------
+Currently, the `Task`  do and undo handlers are started without holding the State lock, but to simplify consistency, it's easier if a `Task` executes while holding the State lock.
+Strictly, `State` lock must only be released when performing slow operations, such as:
+-   copying, compressing or uncompressing of large amounts of on-disk data
+-   network operations
+
+So in practice, most handler code should start with:
+
+```
+st.Lock()
+defer st.Unlock()
+```
+
+where `st` is the runtime `state.State` instance, accessible via `Task.State() or the handler manager.
+
+The deferred `Unlock` will implicitly commit any working state mutations at the end of the handler.
+
+Due to potential restarts, the do or undo handler logic in a Task may be re-executed if it hasn't already completed. This necessitates the following considerations:
+-   on-disk/external state manipulation should be idempotent or equivalent
+-   working state manipulation should either be idempotent or designed to combine working state mutations with setting the next status of the task. This approach currently requires using `Task.SetStatus` before returning form the handler
+
+If slow operations need to be performed, the required `Unlock/Lock` should happen before any working state manipulation.
+
+If the `State` lock is released during a handler, the code needs to consider that other code could have manipulated some relevant working state.There may be also cases where it’s neither possible nor desirable to hold the `State` lock for the entirety of a state manipulation, such as when a manipulation spans multiple subsystems, and so spans multiple tasks. For all such cases, and to simplify reasoning, snapd offers other coordination mechanisms with differing granularity to the `State` lock.
+
+See also the comment in `overlord/snapstate/handlers.go` about state locking.
+
+Conflicts and `Task` precondition blocking
+-------------------------------------------
+At a higher level, it may be appropriate and simpler to manage whether at most one `Change`/sequence of `Task`s is operating on a given snap at a time. As this could stop the system connecting an interface on a snap that’s being removed, for example, or disable service manipulation while its snap is being installed.
```

Comment:
```suggestion
At a higher level, it may be appropriate and simpler to manage whether at most one `Change`/sequence of `Task`s is operating on a given snap at a time. As this could, for example, stop the system connecting an interface on a snap that’s being removed, or disable service manipulation while its snap is being installed.
```